src/Entities/Boat.cpp

```diff
@@ -142,11 +142,14 @@ void cBoat::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 		return;
 	}
 
-	if (IsBlockWater(m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT)))
+	// A real boat floats.
+	// Propel to top water block and sit slightly beneath the waterline:
```

Comment:
Where do we cancel the speed set here - the comment doesn't exactly align with what this does, which is just handling submerged floating. There isn't any concept of a 'top water block' or stopping at the waterline here.

In fact, because you don't check the block above, a boat stuck 'above the waterline' but below another water block above will not be moved by this code.